File: api/app/schemas/finance.py
from typing import Optional, List
from pydantic import BaseModel


class FinanceActualsDashboardResponse(BaseModel):
    actual_id: str
    employee_name: str
    employee_email: Optional[str] = None
    employee_initials: Optional[str] = None
    project_id: str
    project_name: str
    cost_center_id: str
    cost_center_name: str
    year: int
    month: int
    fte_percent: int
    approval_status: str
    current_approval_step: Optional[str]
    current_approver_name: Optional[str]
    approval_instance_id: Optional[str] = None
    current_step_id: Optional[str] = None
    current_approver_object_id: Optional[str] = None
    can_action: bool = False
    can_proxy_approve_step1: bool = False
    step1_id: Optional[str] = None
    is_delegated: bool = False
    delegated_for: Optional[str] = None

class FinanceCostCenterStatsResponse(BaseModel):
    cost_center_id: str
    cost_center_name: str
    demand_fte: float
    supply_fte: float
    actuals_fte: float


class ProjectBreakdownItem(BaseModel):
    project_id: str
    project_name: str
    demand_fte: float
    supply_fte: float = 0.0
    actuals_fte: float


class FinanceEmployeeStatsResponse(BaseModel):
    resource_id: str
    employee_name: str
    demand_fte: float
    supply_fte: float
    actuals_fte: float
    projects: List[ProjectBreakdownItem] = []


class FinanceSettingResponse(BaseModel):
    setting_key: str
    setting_value: str  # caller parses to int/float as needed
    updated_at: Optional[str] = None


class FinanceSettingUpdate(BaseModel):
    setting_value: str


class ConsolidatedCostByProject(BaseModel):
    project_id: str
    project_name: str
    cost_center_id: Optional[str] = None
    cost_center_name: Optional[str] = None
    year: int
    month: int
    demand_cost: int       # planned labor cost in cents
    actuals_cost: int      # actual labor cost in cents
    externals_cost: int    # external contractor cost in cents
    equipment_cost: int    # equipment cost in cents


class ConsolidatedCostResponse(BaseModel):
    data: List[ConsolidatedCostByProject]
    monthly_fte_cost: int  # cents, so frontend can display rate context


class DemandLineDetail(BaseModel):
    resource_name: str
    fte_percent: int
    cost: int  # cents
    project_name: Optional[str] = None  # populated when drilling from cost center


class ActualLineDetail(BaseModel):
    resource_name: str
    fte_percent: int  # actual_fte_percent
    cost: int  # cents
    project_name: Optional[str] = None


class ExternalLineDetail(BaseModel):
    resource_name: Optional[str]
    description: Optional[str]
    notes: Optional[str]
    hours: int
    rate: int       # cents/hr
    total_cost: int  # cents
    project_name: Optional[str] = None


class EquipmentLineDetail(BaseModel):
    description: Optional[str]
    cost: int  # cents
    project_name: Optional[str] = None


class ConsolidatedCostDetail(BaseModel):
    # One of project_id or cost_center_id will be set depending on drill-down mode
    project_id: Optional[str] = None
    project_name: Optional[str] = None
    cost_center_id: Optional[str] = None
    cost_center_name: Optional[str] = None
    year: int
    month: int
    monthly_fte_cost: int
    demand_lines: List[DemandLineDetail]
    actual_lines: List[ActualLineDetail]
    external_lines: List[ExternalLineDetail]
    equipment_lines: List[EquipmentLineDetail]

File: frontend/src/api/finance.ts
import { apiClient } from './client';

export interface CostCenterStats {
  cost_center_id: string;
  cost_center_name: string;
  demand_fte: number;
  supply_fte: number;
  actuals_fte: number;
}

export interface ProjectBreakdown {
  project_id: string;
  project_name: string;
  demand_fte: number;
  supply_fte?: number;
  actuals_fte: number;
}

export interface EmployeeStats {
  resource_id: string;
  employee_name: string;
  employee_email?: string;
  demand_fte: number;
  supply_fte: number;
  actuals_fte: number;
  projects: ProjectBreakdown[];
  cost_center_id?: string;
  cost_center_name?: string;
  employee_initials?: string;
}

export async function getCostCenterStats(year: number, month: number, costCenterId?: string): Promise<CostCenterStats[]> {
  const params = new URLSearchParams({ year: String(year), month: String(month) });
  if (costCenterId) params.append('cost_center_id', costCenterId);
  const res = await apiClient.get<CostCenterStats[]>(`/finance/actuals-vs-plan?${params.toString()}`);
  return res;
}

export async function getEmployeeStats(
  year: number,
  month: number,
  costCenterId?: string,
  projectId?: string
): Promise<EmployeeStats[]> {
  const params = new URLSearchParams({ year: String(year), month: String(month) });
  if (costCenterId) params.append('cost_center_id', costCenterId);
  if (projectId) params.append('project_id', projectId);
  const res = await apiClient.get<EmployeeStats[]>(`/finance/actuals-vs-plan-by-employee?${params.toString()}`);
  return res;
}

export interface FinanceSetting {
  setting_key: string;
  setting_value: string;
  updated_at?: string;
}

export async function getFinanceSetting(key: string): Promise<FinanceSetting> {
  return apiClient.get<FinanceSetting>(`/finance/settings/${key}`);
}

export async function updateFinanceSetting(key: string, value: string): Promise<FinanceSetting> {
  return apiClient.put<FinanceSetting>(`/finance/settings/${key}`, { setting_value: value });
}

export interface ConsolidatedCostRow {
  project_id: string;
  project_name: string;
  cost_center_id: string | null;
  cost_center_name: string | null;
  cost_center_code: string | null;
  year: number;
  month: number;
  demand_cost: number;    // cents — planned labor
  actuals_cost: number;   // cents — actual labor
  externals_cost: number; // cents — external contractors
  equipment_cost: number; // cents — equipment
}

export interface ConsolidatedCostResponse {
  data: ConsolidatedCostRow[];
  monthly_fte_cost: number; // cents
}

export async function getConsolidatedCosts(params?: {
  project_id?: string;
  cost_center_id?: string;
  year?: number;
  month?: number;
  group_by?: 'id' | 'code';
}): Promise<ConsolidatedCostResponse> {
  const query = new URLSearchParams();
  if (params?.project_id) query.append('project_id', params.project_id);
  if (params?.cost_center_id) query.append('cost_center_id', params.cost_center_id);
  if (params?.year != null) query.append('year', String(params.year));
  if (params?.month != null) query.append('month', String(params.month));
  if (params?.group_by) query.append('group_by', params.group_by);
  const qs = query.toString();
  return apiClient.get<ConsolidatedCostResponse>(`/finance/consolidated-costs${qs ? `?${qs}` : ''}`);
}

export interface DemandLineDetail {
  resource_name: string;
  fte_percent: number;
  cost: number;       // cents
  project_name: string | null;
  cost_center_name: string | null;
}

export interface ActualLineDetail {
  resource_name: string;
  fte_percent: number;
  cost: number;       // cents
  project_name: string | null;
  cost_center_name: string | null;
}

export interface ExternalLineDetail {
  resource_name: string | null;
  description: string | null;
  notes: string | null;
  hours: number;
  rate: number;       // cents/hr
  total_cost: number; // cents
  project_name: string | null;
}

export interface EquipmentLineDetail {
  description: string | null;
  cost: number;       // cents
  project_name: string | null;
}

export interface ConsolidatedCostDetail {
  project_id: string | null;
  project_name: string | null;
  cost_center_id: string | null;
  cost_center_name: string | null;
  year: number;
  month: number;
  monthly_fte_cost: number;
  demand_lines: DemandLineDetail[];
  actual_lines: ActualLineDetail[];
  external_lines: ExternalLineDetail[];
  equipment_lines: EquipmentLineDetail[];
}

export async function getConsolidatedCostDetail(params: {
  year?: number;
  month?: number;
  project_id?: string;
  cost_center_id?: string;
  cost_center_code?: string;
}): Promise<ConsolidatedCostDetail[]> {
  const query = new URLSearchParams();
  if (params.year != null) query.append('year', String(params.year));
  if (params.month != null) query.append('month', String(params.month));
  if (params.project_id) query.append('project_id', params.project_id);
  if (params.cost_center_id) query.append('cost_center_id', params.cost_center_id);
  if (params.cost_center_code) query.append('cost_center_code', params.cost_center_code);
  return apiClient.get<ConsolidatedCostDetail[]>(`/finance/consolidated-costs/detail?${query.toString()}`);
}
